index.bs

```diff
@@ -2712,7 +2591,181 @@ partial interface MLGraphBuilder {
+  <summary>
+    The <dfn method for=MLGraphBuilder>expand(|input|, |newShape|)</dfn> method steps are:
+  </summary>
+  <div class=algorithm-steps>
+    <div class="note">
+        The permissions and context validity have been checked by [[#api-mlgraphbuilder-constructor]] steps.
+    </div>
+    1. [=Assert=]: the type of |input| is {{MLOperand}} object.
+    1. [=Assert=]: the type of |newShape| is a `sequence of unsigned long`.
+    1. If any of the following steps fail, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+        1. Let |desc| be |input|.{{MLOperand/[[descriptor]]}}.
+        1. If the sequence length of |newShape| is greater than the [=rank=] of |desc|, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+        1. [=map/For each=] |index| in [=the range=] 0 to the [=rank=] of |input|, exclusive:
+            1. Let |size| be |input|.{{MLOperand/[[descriptor]]}}.{{MLOperandDescriptor/dimensions}}[index].
+            1. If |size| is not equal to 1 and not equal to |newShape|[index], then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+    1. If any of the following sub-steps fail, [=exception/throw=] an "{{OperationError}}" {{DOMException}}.
+        1. Let |output| be the result of <a>creating an MLOperand</a> given [=this=] and |desc|.
+        1. Make a request to the underlying platform to:
+            1. Create an [=implementation-defined=] platform operator |expandImpl| for this method, given |input| and |newShape|.
+            1. Store a reference of |expandImpl| in |output|.{{MLOperand/[[operator]]}}.
+            1. Create an [=implementation-defined=] platform operand |outputImpl| to represent output,given |output| and |expandImpl|.
+            1. Store a reference to |outputImpl| in |output|.{{MLOperand/[[operand]]}}.
+        1. Connect |input| as input to |expandImpl|.
+        1. Connect |output|.{{MLOperand/[[operand]]}} as output to |expandImpl|.
+    1. Return |output|.
+  </div>
+</details>
+
+### gather ### {#api-mlgraphbuilder-gather}
+Gather values of the input tensor along an axis according to the indices.
+<script type=idl>
+dictionary MLGatherOptions {
+  unsigned long axis = 0;
+};
+
+partial interface MLGraphBuilder {
+  MLOperand gather(MLOperand input, MLOperand indices, optional MLGatherOptions options = {});
+};
+</script>
+
+{{MLGatherOptions}} has the following members:
+<dl dfn-type=dict-member dfn-for=MLGatherOptions>
+    : <dfn>axis</dfn>
+    ::
+        An {{unsigned long}} scalar specifying the axis along which the gathered values are obtained. Its value must be in the range [0, N-1] where N is the [=rank=] of the input tensor.
+</dl>
+
+<div>
+    **Arguments:**
+        - *input*: an {{MLOperand}}. The input N-D tensor from which the values are gathered.
+        - *indices*: an {{MLOperand}}. The indices N-D tensor of the input values to gather. The values must be in the range [0, N] where N is the value of the input shape indexed by *options.axis*.
```

Comment:
I suppose the values range should be [0, N-1], and data type should be unsigned integer (uint32)?

Reply:
It seems TF requires the indices value to be within the range [0, params.shape[axis]): https://www.tensorflow.org/api_docs/python/tf/gather

Reply:
I prefer to move ahead with positive indices for now. Negative indexing doubles the amount of out-of-bounds guard logic at the runtime (i.e. more error prone).

FYI, though undocumented, Torch also doesn't support negative index.

Reply:
Some data points of native ML APIs
* DirectML's [Gather](https://learn.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_gather_operator_desc) supports negative indices in range [-N, N)
* CoreML's [Gather](https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#gatherlayerparams) supports negative indices in range [-N, N)
* TOSA's [Gather](https://www.mlplatform.org/tosa/tosa_spec.html#_gather) support indices in range [0, N)

Regarding to Chromium [CL-5017758](https://chromium-review.googlesource.com/c/chromium/src/+/5017758/comment/3439ff1d_f6ff665a/) review, the consensus is to prototype [-N, N) at the current stage.

Reply:
We've previously agreed that the index values to be defined consistently as unsigned ints. I've fixed the typo range to [0, N-1] as @huningxin pointed out.

Reply:
Opened issue https://github.com/webmachinelearning/webnn/issues/484 to track it. Resolve this thread.